ast/src/parsed/mod.rs

```diff
@@ -947,3 +1039,132 @@ pub struct TypedExpression<Ref = NamespacedPolynomialReference, E = Expression<R
     pub e: Expression<Ref>,
     pub type_scheme: Option<TypeScheme<E>>,
 }
+
+#[cfg(test)]
```

Comment:
I'm wondering if it would be better to move the pattern stuff into its own file since it seems to grow quite a bit :)